sdk/core/core-lro/src/poller.ts

```diff
@@ -0,0 +1,163 @@
+      this.stopped = false;
+    }
+    while (!this.isStopped() && !this.isDone()) {
+      await this.poll();
+      await this.delay();
+    }
+  }
+
+  private async pollOnce(options: { abortSignal?: AbortSignalLike } = {}): Promise<void> {
+    const state = this.getOperationState();
+    try {
+      if (!this.isDone()) {
+        this.operation = await this.operation.update({
+          abortSignal: options.abortSignal,
+          fireProgress: this.fireProgress.bind(this)
+        });
+        if (this.isDone() && this.resolve) {
+          this.resolve(state.result);
+        }
+      }
+    } catch (e) {
+      state.error = e;
+      if (this.reject) {
+        this.reject(e);
+      }
+      throw e;
+    }
+  }
+
+  private fireProgress(state: TState): void {
+    for (const callback of this.pollProgressCallbacks) {
+      callback(state);
+    }
+  }
+
+  private async cancelOnce(options: { abortSignal?: AbortSignal } = {}): Promise<void> {
+    this.operation = await this.operation.cancel(options);
+    if (this.reject) {
+      this.reject(new PollerCancelledError("Poller cancelled"));
+    }
+  }
+
+  public poll(options: { abortSignal?: AbortSignal } = {}): Promise<void> {
+    if (!this.pollOncePromise) {
+      this.pollOncePromise = this.pollOnce(options);
+      const clearPollOncePromise = (): void => {
+        this.pollOncePromise = undefined;
+      };
+      this.pollOncePromise.then(clearPollOncePromise, clearPollOncePromise);
+    }
+    return this.pollOncePromise;
+  }
+
+  public async pollUntilDone(): Promise<TResult> {
+    if (this.stopped) {
+      this.startPolling().catch(this.reject);
+    }
+    return this.promise;
+  }
+
+  public onProgress(callback: (state: TState) => void): CancelOnProgress {
+    this.pollProgressCallbacks.push(callback);
+    return (): void => {
+      this.pollProgressCallbacks = this.pollProgressCallbacks.filter((c) => c !== callback);
+    };
+  }
+
+  public isDone(): boolean {
+    const state = this.getOperationState();
+    return Boolean(state.completed || state.cancelled || state.error);
+  }
+
+  public stop(): void {
+    if (!this.stopped) {
+      this.stopped = true;
+      if (this.reject) {
+        this.reject(new PollerStoppedError("This poller is already stopped"));
+      }
+    }
+  }
+
+  public isStopped(): boolean {
+    return this.stopped;
+  }
+
+  public cancelOperation(options: { abortSignal?: AbortSignal } = {}): Promise<void> {
+    if (!this.stopped) {
+      this.stopped = true;
+    }
+    if (!this.cancelPromise) {
+      this.cancelPromise = this.cancelOnce(options);
+    } else if (options.abortSignal) {
+      throw new Error("A cancel request is currently pending");
+    }
+    return this.cancelPromise;
+  }
+
+  public getOperationState(): PollOperationState<TResult> {
+    return this.operation.state;
+  }
+
+  public getResult(): TResult | undefined {
+    if (!this.isDone()) {
+      throw new Error("The poller hasn't finished");
```

Comment:
maybe include a suggestion of checking `isDone()` or using `pollUntilDone()`?

Reply:
For sure! Thank you!